controllers/apim/ratelimitpolicy_controller.go

```diff
@@ -356,105 +355,145 @@ func ratelimitFiltersPatch(gwKey client.ObjectKey, gwLabels map[string]string) (
+		Match: &networkingv1alpha3.EnvoyFilter_EnvoyConfigObjectMatch{
+			Context: networkingv1alpha3.EnvoyFilter_GATEWAY,
+			ObjectTypes: &networkingv1alpha3.EnvoyFilter_EnvoyConfigObjectMatch_RouteConfiguration{
+				RouteConfiguration: &networkingv1alpha3.EnvoyFilter_RouteConfigurationMatch{
+					Vhost: &networkingv1alpha3.EnvoyFilter_RouteConfigurationMatch_VirtualHostMatch{
+						Name: vHostName,
+						Route: &networkingv1alpha3.EnvoyFilter_RouteConfigurationMatch_RouteMatch{
+							Name: "." + routeName, // Istio adds '.' infront of names
+						},
+					},
+				},
 			},
-		}
+		},
+		Patch: patch,
+	}
+}
+
+func virtualHostRateLimitsPatch(vHostName string, rateLimits []*apimv1alpha1.RateLimit) *networkingv1alpha3.EnvoyFilter_EnvoyConfigObjectPatch {
+	// A patch example would look like this:
+	//
+	// applyTo: VIRTUAL_HOST
+	// match:
+	//   context: GATEWAY
+	//   routeConfiguration:
+	//     vhost:
+	// 	    name: api.animaltoys.127.0.0.1.nip.io:80
+	// 	    route:
+	// 	      name: getToys
+	// patch:
+	//   operation: MERGE
+	//   value:
+	//     rate_limits:
+	// 	    - stage: 0
+	// 	      actions:
+	//           - request_headers:
+	//               header_name: ":path"
+	//               descriptor_key: "req.path"
+	//           - request_headers:
+	//               header_name: ":method"
+	//               descriptor_key: "req.method"
+	patchUnstructured := map[string]interface{}{
+		"operation": "MERGE",
+		"value": map[string]interface{}{
+			"rate_limits": common.EnvoyFilterRatelimitsUnstructured(rateLimits),
+		},
+	}
 
-		finalRoute := Patch.Value.Fields["route"]
-		ogRateLimits := finalRoute.GetStructValue().Fields["rate_limits"].GetListValue().Values
-		finalRateLimits := append(ogRateLimits, updatedRateLimits[0])
-		finalRoute.GetStructValue().Fields["rate_limits"].GetListValue().Values = finalRateLimits
+	patchRaw, _ := json.Marshal(patchUnstructured)
+	patch := &networkingv1alpha3.EnvoyFilter_Patch{}
+	if err := patch.UnmarshalJSON(patchRaw); err != nil {
+		//TODO(eguzki): handle error
+		panic(err)
 	}
 
-	envoyFilterPatch := &networkingv1alpha3.EnvoyFilter_EnvoyConfigObjectPatch{
-		ApplyTo: networkingv1alpha3.EnvoyFilter_HTTP_ROUTE,
+	return &networkingv1alpha3.EnvoyFilter_EnvoyConfigObjectPatch{
+		ApplyTo: networkingv1alpha3.EnvoyFilter_VIRTUAL_HOST,
 		Match: &networkingv1alpha3.EnvoyFilter_EnvoyConfigObjectMatch{
 			Context: networkingv1alpha3.EnvoyFilter_GATEWAY,
 			ObjectTypes: &networkingv1alpha3.EnvoyFilter_EnvoyConfigObjectMatch_RouteConfiguration{
 				RouteConfiguration: &networkingv1alpha3.EnvoyFilter_RouteConfigurationMatch{
 					Vhost: &networkingv1alpha3.EnvoyFilter_RouteConfigurationMatch_VirtualHostMatch{
 						Name: vHostName,
 						Route: &networkingv1alpha3.EnvoyFilter_RouteConfigurationMatch_RouteMatch{
-							Name: "." + route.Name, // Istio adds '.' infront of names
+							Action: networkingv1alpha3.EnvoyFilter_RouteConfigurationMatch_RouteMatch_ANY,
```

Comment:
This might be unnecessary I guess

Reply:
If we are applying patch to the virtual host, virtualhost name should be enough for matching.